frontend/server/src/main/java/org/pytorch/serve/util/PriorityLinkedBlockingDeque.java

```diff
@@ -19,70 +20,52 @@
-        for (int priority = 0; priority < this.nPriorities; priority++) {
-            this.priorityDeques.put(priority, new LinkedBlockingDeque<T>(queueSize));
-            if (priority > 0) {
-                // priority weights are inverse priority values for now
-                int priorityWeight = this.nPriorities - priority;
-                for (int key = keyStart; key < keyStart + priorityWeight; key++) {
-                    this.weightedPriorityMap[key] = priority;
-                }
-                keyStart += priorityWeight;
-            }
-        }
+    private int queueSize;
+    private float highPrioProb;
+    private ConcurrentHashMap<String, LinkedBlockingDeque<T>> priorityDeques;
+
+    public PriorityLinkedBlockingDeque(int queueSize, float highPrioProb) {
+
+        this.queueSize = queueSize;
+        this.highPrioProb = highPrioProb;
+        this.priorityDeques = new ConcurrentHashMap<String, LinkedBlockingDeque<T>>();
+
+        // initialize priority deques
+        this.priorityDeques.put("max", new LinkedBlockingDeque<T>(queueSize));
+        this.priorityDeques.put("high", new LinkedBlockingDeque<T>(queueSize));
+        this.priorityDeques.put("low", new LinkedBlockingDeque<T>(queueSize));
+
     }
 
     private LinkedBlockingDeque<T> getDequeForExtraction() {
 
-        // always select deque 0 first if non-empty
-        if (this.nPriorities == 1 || !this.priorityDeques.get(0).isEmpty()) {
-            return this.priorityDeques.get(0);
+        // always select deque max first if non-empty
+        if (!this.priorityDeques.get("max").isEmpty()) {
+            return this.priorityDeques.get("max");
         }
 
-        // sample according to weight map
-        int randInt = ThreadLocalRandom.current().nextInt(sumPriorityWeights);
-        int randPriority = this.weightedPriorityMap[randInt];
-        LinkedBlockingDeque<T> dequeForExtraction = this.priorityDeques.get(randPriority);
-
-        // if sampled deque is empty, scan deques according to priority
-        if (dequeForExtraction.isEmpty()) {
-            for (int priority = 1; priority < this.nPriorities; priority++) {
-                LinkedBlockingDeque<T> priorityDeque = this.priorityDeques.get(priority);
-                if (!priorityDeque.isEmpty()) {
-                    return priorityDeque;
-                }
+        // return random deque unless it is empty and the other is non-empty
+        if (ThreadLocalRandom.current().nextFloat() < this.highPrioProb) {
+            if (this.priorityDeques.get("high").isEmpty() && !this.priorityDeques.get("low").isEmpty()) {
+                return this.priorityDeques.get("low");
+            }
+            return this.priorityDeques.get("high");
+        } else {
+            if (this.priorityDeques.get("low").isEmpty() && !this.priorityDeques.get("high").isEmpty()) {
+                return this.priorityDeques.get("high");
             }
+            return this.priorityDeques.get("low");
```

Comment:
Maybe we could refactor this, so that the (potentially expensive) call to the random number generator is only done, when neither queue is empty.

Reply:
I think if both are empty you also don't have to call the random generator, just return an arbitrary one.

Reply:
Yes, you're right. I initially thought it did matter because the worker was listening for that specific deque, but in the current implementation it will call `getDequeForExtraction()` again anyway whenever `notEmpty` is signalled.